clients/src/main/java/org/apache/kafka/common/serialization/ListDeserializer.java

```diff
@@ -0,0 +1,196 @@
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
+
+public class ListDeserializer<Inner> implements Deserializer<List<Inner>> {
+
+    private Deserializer<Inner> inner;
+    private Class<?> listClass;
+    private Integer primitiveSize;
+
+    static private Map<Class<? extends Deserializer<?>>, Integer> fixedLengthDeserializers = mkMap(
+        mkEntry(ShortDeserializer.class, 2),
+        mkEntry(IntegerDeserializer.class, 4),
+        mkEntry(FloatDeserializer.class, 4),
+        mkEntry(LongDeserializer.class, 8),
+        mkEntry(DoubleDeserializer.class, 8),
+        mkEntry(UUIDDeserializer.class, 36)
+    );
+
+    public ListDeserializer() {}
+
+    public <L extends List<Inner>> ListDeserializer(Class<L> listClass, Deserializer<Inner> innerDeserializer) {
+        this.listClass = listClass;
+        this.inner = innerDeserializer;
+        if (innerDeserializer != null) {
+            this.primitiveSize = fixedLengthDeserializers.get(innerDeserializer.getClass());
+        }
+    }
+
+    public Deserializer<Inner> getInnerDeserializer() {
+        return inner;
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        if (listClass == null) {
+            configureListClass(configs, isKey);
+        }
+        if (inner == null) {
+            configureInnerSerde(configs, isKey);
+        }
+    }
+
+    private void configureListClass(Map<String, ?> configs, boolean isKey) {
+        String listTypePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_TYPE_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_TYPE_CLASS;
+        final Object listClassOrName = configs.get(listTypePropertyName);
+        if (listClassOrName == null) {
+            throw new ConfigException("Not able to determine the list class because it was neither passed via the constructor nor set in the config.");
+        }
+        try {
+            if (listClassOrName instanceof String) {
+                listClass = Utils.loadClass((String) listClassOrName, Object.class);
+            } else if (listClassOrName instanceof Class) {
+                listClass = (Class<?>) listClassOrName;
+            } else {
+                throw new KafkaException("Could not determine the list class instance using \"" + listTypePropertyName + "\" property.");
+            }
+        } catch (final ClassNotFoundException e) {
+            throw new ConfigException(listTypePropertyName, listClassOrName, "Deserializer's list class \"" + listClassOrName + "\" could not be found.");
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private void configureInnerSerde(Map<String, ?> configs, boolean isKey) {
+        String innerSerdePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_INNER_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_INNER_CLASS;
```

Comment:
What about the `LIST_KEY_DESERIALIZER_INNER_CLASS_CONFIG`?

Reply:
It was introduced in https://github.com/apache/kafka/pull/6592/commits/f91d75925aac4dce1f9522b45142ea25b5258748
However, now I am looking at it and seems like we actually don't need either of:
```
LIST_KEY_DESERIALIZER_INNER_CLASS_CONFIG
LIST_KEY_DESERIALIZER_TYPE_CLASS_CONFIG
LIST_KEY_SERIALIZER_INNER_CLASS_CONFIG
LIST_KEY_SERIALIZER_TYPE_CLASS_CONFIG
```

Since we are operating only with:
```
DEFAULT_LIST_KEY_SERDE_INNER_CLASS
DEFAULT_LIST_KEY_SERDE_TYPE_CLASS
DEFAULT_LIST_VALUE_SERDE_INNER_CLASS
DEFAULT_LIST_VALUE_SERDE_TYPE_CLASS
```

Reply:
Cool. I think the fewer configs overall, the better. If we can get away with just the Serde configs then let's do so to keep the API surface area smaller for users 👍